### This script includes the following points:

#### 1. Import the data
#### 2. Wrangle the data
#### 3. Finding Missing values
#### 4. Finding Duplicates
#### 5. Finding Mixed-Type Data in any column
#### 6. Combining two dataframes
#### 7. Exporting DataFrames

### 1. Impport the data

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
path = r"C:\Users\Poory\OneDrive\Desktop\project\Instacart Basket Analysis"

In [3]:
df_cust = pd.read_csv(os.path.join(path,"02 Data", "Original Data", "customers.csv"), index_col = False)
df_main = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_tsk48.pkl')) 

In [4]:
df_cust

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374  
...              ...        ...     ...  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095  

[206209 rows x 10 columns]

In [5]:
df_cust.shape

(206209, 10)

In [6]:
df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [7]:
df_cust.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

### 2. Wrangle the data

In [8]:
# Make all  columns lower case
df_cust.columns.str.lower()

Index(['user_id', 'first name', 'surnam', 'gender', 'state', 'age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')

In [9]:
df_cust.columns = df_cust.columns.str.lower()

In [10]:
df_cust

user_id first name    surnam  gender           state  age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374  
...              ...        ...     ...  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095  

[206209 rows x 10 columns]

In [11]:
# rename two columns
df_cust.rename(columns = {'first name':'first_name', 'surnam':'last_name'}, inplace=True)

### 3. Finding Missing values

In [12]:
df_cust.isnull().sum()

user_id             0
first_name      11259
last_name           0
gender              0
state               0
age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

In [13]:
df_nan = df_cust.loc[df_cust['first_name'].isnull() == True]
df_nan

user_id first_name last_name  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        n_dependants fam_status  income  
53                 2    married   41709  
73                 0     single   82518  
82                 3    married  117099  
99                 1    married  155673  
105                3    married  151819  
...              ...        ...     ...  
206038             3    married   87783  
206044             2    married  108488  
206090             1    married   45275  
206162             0     single   56325  
206171             2    married   59222  

[11259 rows x 10 columns]

#### It seems we have 11259 rows with missing values in the first_name column. However, in my opinion, we have user_id and customers' last names with their age and state, which helps us differentiate between two people with the same last name. On the other hand, I don't think we should delete these rows, although they are a small proportion of the entire data set. So, I think we can just ignore this column because we don't need it for our analysis.

In [14]:
df = df_cust.drop(columns= 'first_name')

#### 4. Finding Duplicates

In [15]:
df_dup = df[df.duplicated()]
df_dup

Empty DataFrame
Columns: [user_id, last_name, gender, state, age, date_joined, n_dependants, fam_status, income]
Index: []

In [16]:
# Deleting duplicated values
df = df.drop_duplicates()

In [17]:
df.isnull().sum()

user_id         0
last_name       0
gender          0
state           0
age             0
date_joined     0
n_dependants    0
fam_status      0
income          0
dtype: int64

In [18]:
# I want to check if there are any misspellings in states or marriage status
df['state'].value_counts()

state
Florida                 4044
Colorado                4044
Illinois                4044
Alabama                 4044
District of Columbia    4044
Hawaii                  4044
Arizona                 4044
Connecticut             4044
California              4044
Indiana                 4044
Arkansas                4044
Alaska                  4044
Delaware                4044
Iowa                    4044
Idaho                   4044
Georgia                 4044
Wyoming                 4043
Mississippi             4043
Oklahoma                4043
Utah                    4043
New Hampshire           4043
Kentucky                4043
Maryland                4043
Rhode Island            4043
Massachusetts           4043
Michigan                4043
New Jersey              4043
Kansas                  4043
South Dakota            4043
Minnesota               4043
Tennessee               4043
New York                4043
Washington              4043
Louisiana               4043
Montana 

In [19]:
df['fam_status'].value_counts()

fam_status
married                             144906
single                               33962
divorced/widowed                     17640
living with parents and siblings      9701
Name: count, dtype: int64

In [20]:
df['gender'].value_counts()

gender
Male      104067
Female    102142
Name: count, dtype: int64

### 5. Finding Mixed-Type Data in any column

In [21]:
# Check for columns with inconsistent data types in specific column (chatGPT version)
for col in df.columns.tolist():
    
    # Check if there are any inconsistencies in data types within the column
    weird = (df[[col]].apply(lambda x: x.map(type)) != df[[col]].iloc[0].apply(type)).any(axis=1)
    
    # If inconsistencies are found, print the column name
    if len(df[weird]) > 0:
        print(col)

In [22]:
df.shape

(206209, 9)

In [23]:
df_main.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   last_order_days  first_order  product_id  add_to_cart_order  reordered  \
0              NaN         True         196                  1          0   
1             15.0        False         196                  1          1   
2             21.0        False         196                  1          1   
3             29.0        False         196                  1          1   
4             28.0        False         196                  1          1   

   ...        price_range     busiest_day         busiest_days  \
0  ...  Mid-range product  Regularly busy       Regularly busy   
1  ...  Mid-range product  Regularly busy  Two Least busy days   
2  ...  Mid-range product  Regularly busy  Two Least busy days   
3  ...  Mid-range product      Least busy  Two Least busy days   
4  ...  Mid-range product      Least busy  Two Least busy days   

   busiest_period_of_day max_order  loyalty_flag avg_spent spending_flag  \
0         Average Orders        10  New customer  6.367797   Low spender   
1         Average Orders        10  New customer  6.367797   Low spender   
2            Most Orders        10  New customer  6.367797   Low spender   
3         Average Orders        10  New customer  6.367797   Low spender   
4            Most Orders        10  New customer  6.367797   Low spender   

  order_freq_med         frequency_flag  
0           20.5  Non-frequent customer  
1           20.5  Non-frequent customer  
2           20.5  Non-frequent customer  
3           20.5  Non-frequent customer  
4           20.5  Non-frequent customer  

[5 rows x 25 columns]

In [24]:
pd.options.display.max_columns = None

In [25]:
df_main.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   last_order_days  first_order  product_id  add_to_cart_order  reordered  \
0              NaN         True         196                  1          0   
1             15.0        False         196                  1          1   
2             21.0        False         196                  1          1   
3             29.0        False         196                  1          1   
4             28.0        False         196                  1          1   

  product_name  aisle_id  department_id  prices _merge        price_range  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day         busiest_days busiest_period_of_day  max_order  \
0  Regularly busy       Regularly busy        Average Orders         10   
1  Regularly busy  Two Least busy days        Average Orders         10   
2  Regularly busy  Two Least busy days           Most Orders         10   
3      Least busy  Two Least busy days        Average Orders         10   
4      Least busy  Two Least busy days           Most Orders         10   

   loyalty_flag  avg_spent spending_flag  order_freq_med  \
0  New customer   6.367797   Low spender            20.5   
1  New customer   6.367797   Low spender            20.5   
2  New customer   6.367797   Low spender            20.5   
3  New customer   6.367797   Low spender            20.5   
4  New customer   6.367797   Low spender            20.5   

          frequency_flag  
0  Non-frequent customer  
1  Non-frequent customer  
2  Non-frequent customer  
3  Non-frequent customer  
4  Non-frequent customer

In [26]:
df_main['user_id'].dtype

dtype('int32')

In [27]:
df['user_id'].dtype

dtype('int64')

In [28]:
df['user_id'] = df['user_id'].astype('int32')

In [29]:
df['user_id'].dtype

dtype('int32')

### 6. Combining two dataframes

In [30]:
df_main.drop(columns = '_merge', inplace = True)

In [31]:
df_merged = df_main.merge(df, on = 'user_id', indicator = True)

In [36]:
df_merged['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

In [39]:
df_merged.drop(columns = '_merge', inplace = True)

In [40]:
df_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   last_order_days  first_order  product_id  add_to_cart_order  reordered  \
0              NaN         True         196                  1          0   
1             15.0        False         196                  1          1   
2             21.0        False         196                  1          1   
3             29.0        False         196                  1          1   
4             28.0        False         196                  1          1   

  product_name  aisle_id  department_id  prices        price_range  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day         busiest_days busiest_period_of_day  max_order  \
0  Regularly busy       Regularly busy        Average Orders         10   
1  Regularly busy  Two Least busy days        Average Orders         10   
2  Regularly busy  Two Least busy days           Most Orders         10   
3      Least busy  Two Least busy days        Average Orders         10   
4      Least busy  Two Least busy days           Most Orders         10   

   loyalty_flag  avg_spent spending_flag  order_freq_med  \
0  New customer   6.367797   Low spender            20.5   
1  New customer   6.367797   Low spender            20.5   
2  New customer   6.367797   Low spender            20.5   
3  New customer   6.367797   Low spender            20.5   
4  New customer   6.367797   Low spender            20.5   

          frequency_flag last_name  gender    state  age date_joined  \
0  Non-frequent customer    Nguyen  Female  Alabama   31   2/17/2019   
1  Non-frequent customer    Nguyen  Female  Alabama   31   2/17/2019   
2  Non-frequent customer    Nguyen  Female  Alabama   31   2/17/2019   
3  Non-frequent customer    Nguyen  Female  Alabama   31   2/17/2019   
4  Non-frequent customer    Nguyen  Female  Alabama   31   2/17/2019   

   n_dependants fam_status  income  
0             3    married   40423  
1             3    married   40423  
2             3    married   40423  
3             3    married   40423  
4             3    married   40423

In [41]:
df_merged.shape

(32404859, 32)

### 7. Exporting DataFrames

In [42]:
df_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'final.pkl'))